File: templates/matlab/bump_chart.m
function fig = bump_chart()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(11);
    n_item = 6; n_per = 10;
    scores = cumsum(randn(n_item, n_per), 2);
    ranks = zeros(n_item, n_per);              % rank 1 = highest score
    for j = 1:n_per
        [~, ord] = sort(scores(:, j), 'descend');
        ranks(ord, j) = 1:n_item;
    end
    labels = {'Alpha', 'Bravo', 'Charlie', 'Delta', 'Echo', 'Foxtrot'};
    t_fine = linspace(0, n_per-1, 200);
    seg = min(max(floor(t_fine), 0), n_per-2);
    frac = t_fine - seg;
    ease = 3*frac.^2 - 2*frac.^3;              % smoothstep easing
    fig = figure('Position', [100 100 700 400]); hold on;
    for i = 1:n_item
        y = ranks(i, seg+1) + (ranks(i, seg+2) - ranks(i, seg+1)).*ease;
        c = palette('cat', i);
        plot(t_fine, y, 'Color', c, 'LineWidth', 2);
        plot(0:n_per-1, ranks(i, :), 'o', 'Color', c, 'MarkerFaceColor', c, ...
             'MarkerSize', 5);
        text(n_per-0.7, ranks(i, end), labels{i}, 'Color', c, 'FontSize', 8, ...
             'VerticalAlignment', 'middle');
    end
    xlim([-0.3, n_per+1.0]); ylim([0.5, n_item+0.5]);
    set(gca, 'YDir', 'reverse', 'YTick', 1:n_item, 'XTick', 0:n_per-1, ...
        'XTickLabel', arrayfun(@(q) sprintf('Q%d', q), 1:n_per, 'UniformOutput', false));
    xlabel('period'); ylabel('rank'); title('Ranking over time (bump chart)');
    ax = gca; ax.XGrid = 'on'; ax.YGrid = 'off';
end

File: templates/python/bump_chart.py
"""bump_chart: 排名随时间变化的凹凸图（6 项目 x 10 期，平滑过渡）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(title='Ranking over time (bump chart)'):
    apply_theme(fig_size=(7, 4))
    rng = np.random.default_rng(11)
    n_item, n_per = 6, 10
    scores = np.cumsum(rng.normal(0, 1, (n_item, n_per)), axis=1)
    ranks = scores.shape[0] - np.argsort(np.argsort(scores, axis=0), axis=0)
    labels = ['Alpha', 'Bravo', 'Charlie', 'Delta', 'Echo', 'Foxtrot']
    t_fine = np.linspace(0, n_per - 1, 200)
    seg = np.clip(np.floor(t_fine).astype(int), 0, n_per - 2)
    frac = t_fine - seg
    ease = 3 * frac**2 - 2 * frac**3          # smoothstep easing
    fig, ax = plt.subplots()
    for i in range(n_item):
        y = ranks[i, seg] + (ranks[i, seg + 1] - ranks[i, seg]) * ease
        ax.plot(t_fine, y, color=cycle(i), linewidth=2)
        ax.plot(np.arange(n_per), ranks[i], 'o', color=cycle(i), markersize=5)
        ax.text(n_per - 0.7, ranks[i, -1], labels[i], color=cycle(i),
                va='center', fontsize=8)
    ax.set_xlim(-0.3, n_per + 1.0)
    ax.invert_yaxis()
    ax.set_yticks(np.arange(1, n_item + 1))
    ax.set_xticks(np.arange(n_per))
    ax.set_xticklabels([f'Q{q+1}' for q in range(n_per)])
    ax.set_xlabel('period'); ax.set_ylabel('rank'); ax.set_title(title)
    ax.grid(True, axis='x', linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)
